src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,441 @@
+        """Identify the format of the files in the dictionary."""
+        union = set()
+        identified_formats_by_name = {
+            k: identify_file_format(k) for k, v in self.root.items()
+        }
+        stored_formats_by_name = {
+            k: v.format for k, v in self.root.items() if hasattr(v, "format")
+        }
+        if not all(
+            k in identified_formats_by_name and identified_formats_by_name[k] == v
+            for k, v in stored_formats_by_name.items()
+        ):
+            raise ValueError(
+                f"identified formats and stored formats do not match: identified formats: {identified_formats_by_name}, stored formats: {stored_formats_by_name}"
+            )
+        union.update(identified_formats_by_name.values())
+        if len(union) == 1:
+            return union.pop()
+        return "|".join(union)
+
+    @model_validator(mode="before")
+    def preproc_data(cls, data: Any) -> Any:
+        data = copy.deepcopy(data)
+        for k, v in data.items():
+            if isinstance(v, (str, type(None))):
+                data[k] = {"object_path": v}
+                v = data[k]
+            if isinstance(v, dict):
+                fmt = identify_file_format(k)
+                if fmt == "root":
+                    if "format" not in v:
+                        v["format"] = "root"
+                    else:
+                        assert (
+                            v["format"] == "root"
+                        ), f"Expected 'format' to be 'root', got {v['format']} for {k}"
+                elif fmt == "parquet":
+                    if "format" not in v:
+                        v["format"] = "parquet"
+                    else:
+                        assert (
+                            v["format"] == "parquet"
+                        ), f"Expected 'format' to be 'parquet', got {v['format']} for {k}"
+        return data
+
+
+class InputFiles(
+    RootModel[
+        dict[
+            str,
+            Union[
+                CoffeaROOTFileSpec,
+                CoffeaParquetFileSpec,
+                CoffeaROOTFileSpecOptional,
+                CoffeaParquetFileSpecOptional,
+            ],
+        ]
+    ],
+    DictMethodsMixin,
+    InputFilesMixin,
+):
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.root)
+
+    @model_validator(mode="after")
+    def promote_and_check_files(self) -> Self:
+        preprocessed = {k: False for k in self.root}
+        for k, v in self.root.items():
+            try:
+                if type(v) in [CoffeaROOTFileSpecOptional]:
+                    self.root[k] = CoffeaROOTFileSpec(v)
+                if type(v) in [CoffeaParquetFileSpecOptional]:
+                    self.root[k] = CoffeaParquetFileSpec(v)
+                preprocessed[k] = True
+            except Exception:
+                pass
+        if all(preprocessed.values()):
+            try:
+                self.root = PreprocessedFiles(self.root).root
```

Comment:
This check seems redundant with the above. Anyway, is there any harm in letting `InputFiles` always hold optional FileSpec types?